### PR TITLE
Plumbing chemical chamber UI and operation improvements

### LINKED ISSUE #58119: Reaction Chambers do not accept pH buffers
## Testmerges:

#57405 : 'Being shot into space by the mass driver now reconnects you to Terry (SHOULD WORK NOW)' by Ryll-Ryll at commit 1eb7b0169c
## Reproduction:

Assemble a chemical factory such that you are supplying acidic and basic buffer on layer 2, yellow and green pipes respectively, and set a reaction chamber with an appropriate pH range to produce a chemical like Libital. The reaction chamber will accept pH buffer once, and only once after being constructed (you will see a message in chat from it). Thereafter the reaction chamber will not take any buffer from the green and yellow duct networks, but the chamber will continue to cycle without adjusting its pH.

Unanchoring and re-anchoring the reactor does not restore the behavior, nor does plungering it.

In the screenshot below I would expect the reaction chamber to keep the contents within a range of 9-10 for the libital reaction. Instead it plunges to 1. Changing the values in the limits appears to have no impact on the pH of the reaction or the purity of the products.

imageimage

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
1. Fixes #58119

   The word "Fixes" actually means the reaction chamber does it's job but "**does not guarantee**" the ph & temperature of the final solution is exactly within the ranges you specify especially for instantaneous reactions. There were 2 problems with the reaction chamber.

    a) **The ph chemistry of the reaction chamber was completely flipped upside down.** 
      Let's look at the code
       https://github.com/tgstation/tgstation/blob/cfe13015203480dafaf3fce1589ad6045e708442/code/modules/plumbing/plumbers/reaction_chamber.dm#L152-L155

   This was adding basic buffer when the solution was getting to basic & acid buffer when the solution was getting too acidic

      The correct mode of operation is as follows
    - If the ph is getting way too **acidic** i.e. it's ph **decreases** and becomes **lower** than the acidic limit then we need to add some **base buffer** so as to **increase** it's value and bring it back **above** the acidic limit
    - If the ph is getting way to **basic** i.e. it's ph **increases** and becomes **higher** than the alkaline limit then we need to add some **acidic buffer** so as to **decrease** it's value and bring it back **below** the alkaline limit

    b) **The reaction chamber did it's work only half of the time**
   Looking at the code
  https://github.com/tgstation/tgstation/blob/cfe13015203480dafaf3fce1589ad6045e708442/code/modules/plumbing/plumbers/reaction_chamber.dm#L152

    The reaction chamber would only balance the ph of the solution **ONLY** when an reaction was taking place i.e. only when `reagents.is_reacting` is true. It would not attempt to balance the ph of the reagents as and when they were coming in i.e when `emptying` is FALSE. This means if an reaction took only like 0.5 to 1 second to occur the ph would be balanced only during that very short interval of time and so it would not make much of difference.

    So the 2 problems were the chamber was doing its job wrong and would do it only half the time. Again re-emphasizing even with these patches the chamber "**does not guarantee**" the final solution ph will be within those ranges. This is because the nature of reactions is unpredictable. Some reactions can take place so fast and have huge drastic ph changes that the reaction chamber would be ineffective regardless of how much base/acid it used to balance the reaction. The only thing is that it does it's job correctly and too the best of its ability with no guarentees.

3. The plumbing mixing chamber(parent type of the plumbing chemical reaction chamber) which you make via research plumbing RCD, UI now also uses an TGUI input list to accept chemicals so no more manual typing.

4. Both these types have their UI files converted Typescript cause why not

## Changelog
:cl:
fix: plumbing reaction chamber now balances the ph of it's solution correctly to the best of it's ability so no guarantees
code: converted plumbing reaction chamber & mixing chamber UI files to Typescript
refactor: plumbing mixing chamber now also accepts an TGUI input list to input it's chemicals
/:cl:

### CODE REVIEW

**Comment by SyncIt21** [comment]
@Time-Green so umm.. you free? This is your review after all :)

File: code/modules/plumbing/plumbers/reaction_chamber.dm [hunk 1]
```diff
@@ -147,13 +172,31 @@
...
+			return
+
+		//transfer buffer and handle reactions, not a proven math but looks logical
+		var/transfer_amount = FLOOR((reagents.ph > alkaline_limit ? (reagents.ph - alkaline_limit) : (acidic_limit - reagents.ph)) * seconds_per_tick, CHEMICAL_QUANTISATION_LEVEL)
+		if(transfer_amount <= 0 || !buffer.trans_to(reagents, transfer_amount))
+			return
+		reagents.handle_reactions()
+
+		//some power for accurate ph balancing
+		use_power((active_power_usage / 5) * seconds_per_tick)
```
Comment (Time-Green) [inline code/modules/plumbing/plumbers/reaction_chamber.dm:200]: Can you restructure this to not use a `while(TRUE)`? 
You can definitely just do this with a regular while()

File: code/modules/plumbing/plumbers/reaction_chamber.dm [hunk 2]
```diff
@@ -147,13 +172,31 @@
...
```
Comment (Time-Green) [inline code/modules/plumbing/plumbers/reaction_chamber.dm:199]:
```suggestion
		use_power((active_power_usage * 0.2) * seconds_per_tick)
```
Multiplication is preferred if possible

**Review by Time-Green — CHANGES_REQUESTED** [review]
Thingies

Also I dont understand UI but it looks fine I think

File: code/modules/plumbing/plumbers/reaction_chamber.dm [hunk 3]
```diff
@@ -35,30 +35,48 @@
...
+		//adjust temperature of final solution
+		var/temp_diff = target_temperature - reagents.chem_temp
+		if(abs(temp_diff) > 0.01) //if we are not close enough keep going
+			reagents.adjust_thermal_energy(temp_diff * HEATER_COEFFICIENT * seconds_per_tick * SPECIFIC_HEAT_DEFAULT * reagents.total_volume) //keep constant with chem heater
+
+		//do other stuff with final solution
+		handle_reagents(seconds_per_tick)
+
+		//full power for doing reactions
+		power_usage *= 2
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/plumbing/plumbers/reaction_chamber.dm:68]: But that's what I am doing right? It's half outside the if statement(when stuff is not reacting) and becomes full value inside it(when stuff is reacting).

File: code/modules/plumbing/plumbers/reaction_chamber.dm [hunk 4]
```diff
@@ -104,32 +120,41 @@
...
-			. = TRUE
+			return TRUE
 
 		if("temperature")
 			var/target = text2num(params["target"])
 			if(target != null)
-				target_temperature=clamp(target, 0, 1000)
-			.=TRUE
+				target_temperature = clamp(target, 0, 1000)
+			return TRUE
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/plumbing/plumbers/reaction_chamber.dm:135]: The UI already has limits in it's number input box from 0 to 1000 but being extra safe also helps.  I hope

File: code/modules/plumbing/plumbers/reaction_chamber.dm (hunk 1 above)
Comment (SyncIt21) [reply code/modules/plumbing/plumbers/reaction_chamber.dm:200]: I could move the ph range check to the while condition sure

File: code/modules/plumbing/plumbers/reaction_chamber.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (Time-Green) [reply code/modules/plumbing/plumbers/reaction_chamber.dm:68]: Yeah I meant as in it looks a bit awkard, but it's moot tbh

**Pull request merged**